src/python/database.py

```
@@ -589,3 +574,42 @@
+            if connection:
+                connection.close()
+
+    def _get_connection(self) -> sqlite3.Connection:
+        """Get a database connection with row factory"""
+        try:
+            conn = sqlite3.connect(self.db_file)
+            conn.row_factory = sqlite3.Row
+            return conn
+        except sqlite3.Error as e:
+            log.error("Failed to connect to database: %s", str(e))
+            raise DatabaseError(f"Database connection failed: {str(e)}", "DB_CONN_ERROR")
```

Ruff (0.8.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)